[New Feature] Sharding / Parallel Execution

sdfgsdfgd opened the pull request:
## Highlights

### Parallel Execution and Sharding
- New shards parameter `--shards 5` or `-s 5` to distribute tests across multiple devices
- Enhance `--deviceid` parameter to allow specification of multiple devices like `--deviceid 'emulator-5554,emulator-5556,1C011FDEE052JE'`
- Implement sharding to distribute test flows evenly across emulators and devices ( theoretically limited by adb port limitation of 64 pairs in one instance of maestro )
- Uses specified devices, if not uses available/connected devices, if not prompts the user for device creation for the missing devices up to shard count
- Wait until all shards are initialised
- Can run even multiple sharded maestro instances of real devices and emulators simultaneously at the same time, up to infinity (or 64? who knows), as much as the host memory allows (Haven't been able to test beyond 7 shards yet, if there is an inefficiency I'll be happy to work on it to improve this so we can indeed support as many devices as possible)
- Improved Port Management and thread synchronizations, use of a cute little CountDownLatch to make shards wait
- Set dynamic port allocation for the AndroidDriver to enhance RPC communication
- JVM Coroutines (version 1.6.3) to improve asynchronous task management
- Small clean up for assemble gradle task by: Disable lint checks on assemble tasks to streamline the build process and quiet the errors

- Enables dynamic reception of the server port parameter from the Activity Manager (am - nifty little tool !)
https://developer.android.com/tools/adb#am
https://developer.android.com/studio/test/command-line


Impact:
these enhancements not only increase the robustness and flexibility of Maestro but also set the stage for future extensions and improvements

Run multiple sharded maestro clusters on a horde of real devices and simulators mixed together, run 15 real devices alongside 17 emulators, as 5 different clusters of 120 different devices - go crazy. Help us find a bottleneck so we can actually achieve this 100% smoothly and reliably. I haven't been able to test beyond almost 5 to 7 shards on my device yet.

### Attachments
demo setup:   4 devices/shards 8 test files in folder, (the tests scroll & tap around settings app)
1. 1 x my real device (scrcpy) 
2. 2 x API 34 android emulators  
3. 1 x prompt created android 34

demo command, run installDist executable, or with run gradle task:
1. `./maestro --device 'emulator-5554,emulator-5556,emulator-5558,1C011FDEE002JE' test -s 5 ~/Desktop `
2. run gradle task `run --args="--device 'emulator-5554,emulator-5556,emulator-5558,1C011FDEE002JE' test -s 5 ~/Desktop`


https://github.com/mobile-dev-inc/maestro/assets/16299826/813666a0-7cf0-4df0-b713-7380b33c9e76


### New different ways to run : 
- `maestro test -s 4 ~/Desktop` 
run with 4 shards on all yaml files under Desktop. Either reuses all online devices or prompts for creation of 4 devices if no running devices found
- `maestro test -s 4 ~/Desktop  --device '1C011FDEE002JE'`
reuses a physical device with serial `1C011FDEE002JE` and either reuses 3 other online devices, or prompts for creation of enough devices to reach 4 shards

axelniklasson commented:
Thanks for taking the time to create this PR @sdfgsdfgd, this will be a great addition! Someone from the team will review it next week and work with you to get this merged.

sdfgsdfgd commented:
thank you, in the meantime

1. noticed a minor bug with device creation's 2nd and later runs if you do sharding without any online devices
2. report file generations for each shard I totally forgot about this, already done with the solution - I'll push them soon

igorsmotto commented on maestro-cli/src/main/java/maestro/cli/command/TestCommand.kt:
```diff
@@ -119,70 +141,186 @@ class TestCommand : Callable<Int> {
+                                reportOut = format.fileExtension
+                                    ?.let { extension ->
+                                        (output ?: File(
+                                            "report${
+                                                if (shardIndex >= 1) {
+                                                    "_" + shardIndex + 1
+                                                } else {
+                                                    ""
+                                                }
```
As a user I feel having to deal with many junit xml files one for each shard is unexpected. Since maestro is handling the parallel execution. I'd expect that we combine them into a single file, given that they all share the same test suite name.

sdfgsdfgd replied:
Sure, I'll address this soon, thank you 👍

igorsmotto commented on maestro-cli/src/main/java/maestro/cli/App.kt:
```diff
@@ -63,10 +64,13 @@ class App {
     var port: Int? = null
 
-    @Option(names = ["--device", "--udid"], description = ["(Optional) Select a device to run on explicitly"])
+    @Option(names = ["--device", "--udid"], description = ["(Optional) Device ID to run on explicitly, can be a comma separated list of IDs: --udid \"Emulator_1,Emulator_2\" "])
```
udid is specifically used because its the term that Apple uses it, so the example being Emulators (as opposed to Simulators) is unexpected

```suggestion
    @Option(names = ["--device", "--udid"], description = ["(Optional) Device ID to run on explicitly, can be a comma separated list of IDs: --device \"Emulator_1,Emulator_2\" "])
```

igorsmotto commented on maestro-cli/src/main/java/maestro/cli/device/DeviceCreateUtil.kt:
```diff
@@ -103,11 +105,11 @@ internal object DeviceCreateUtil {
         // check connected device
-        if (DeviceService.isDeviceConnected(deviceName, Platform.ANDROID) != null) {
-            throw CliError("A device with name $deviceName is already connected")
+        while (DeviceService.isDeviceConnected(deviceName, Platform.ANDROID) != null) {
+            deviceName += "_"
```
Did not follow here, why adding an _ to the name of the device would make this work?  
Maybe related to this line: https://github.com/mobile-dev-inc/maestro/pull/1732/files#diff-cb7f7f66181dc16c2226f299c8649e8f30be45c7c13deeda207d0d255cb9c61cR30?

sdfgsdfgd replied:
you're right, I was taking care of this part around there, where you included the link. 

But I vaguely remember running 2 maestro instances with multiple shards or some weird setup like that, and the other maestro session's first shard had the exact same pixel name, and this part of the code prevented me from running a 2nd maestro with 2 shards. So one with 2 shards, another with 2 shards, second one ends up trying to create device with same name and causes to fail. But this quick workaround lets you create duplicates of shards as well. It might be redundant, maybe sharding is enough by itself :D I'll remove this if you want ?

igorsmotto reviewed (changes requested):
This PR is a big one, but most importantly is a hard one. Your ideas on how to solve this challenge are really good.

Unfortunately, let me say upfront that it will take me a while to fully review it, since of its high complexity.

In the meantime I'll be posting here all my questions as I spend more and more time in this PR.

---
It seems like this whole solution doesn't work on iOS::
1. It can only create one device, probably because it tries to create them all of the same name (Error: `A device with name Maestro_iPhone11_17 is already connected`).
4. Even if I use already created multiple devices, it seems like only one is actually being operated by Maestro, unsure what is going on. 
5. Haven't tested: Unsure what will happen if the user pass a list of devices that has a mix of android and devices in it. We should probably error it out, this definitely means that the user is making a mistake. This is also a weird user experience if no devices is passed, because the user is prompted multiple times what platforms it wants to run that shard.

Is this something planned? To only add support for Android? If so, then we need to error out and properly document that this feature can only work for Android

---
Having shards silently overrides the [sequence execution](https://maestro.mobile.dev/cli/test-suites-and-reports#sequential-execution), which can be unexpected to a user given that the sequence is a workspace-configuration.
Adding support to it is non-trivial, as it would essentially means there is a need to build a graph to have some level of parallel execution, so I'd expect this to error out.

--- 
The TestCommand is responsible for creating those devices, but this suggests that the StartDeviceCommand should also have a shard parameter. This way a user can avoid having to interact with Maestro and just create a bunch of devices in a single sweep.

sdfgsdfgd commented on maestro-android/src/androidTest/java/dev/mobile/maestro/MaestroDriverService.kt:
```diff
@@ -76,7 +85,11 @@ class MaestroDriverService {
         val uiAutomation = instrumentation.uiAutomation
 
-        NettyServerBuilder.forPort(7001)
+        val port = InstrumentationRegistry.getArguments().getString("port", "9001").toInt()
+
+        println("Server running on port [ $port ]")
```
let me explain - the number 7001 or 9001 really doesn't matter any more

this is actually now set by the parameter which gets passed down from default CLI parameters, unless it's specified. A random port between 9001 and 9128 is selected at runtime. 

Below addition from App.kt ;
```
@Option(names = ["--driver-host-port"], hidden = true)
    var driverHostPort: Int = Random.nextInt(9001, 9128)
```

This allows multiple maestro executables to run simultaneously. It's also a required key part of bringing up more than one device. 

Why 128 ? That is because adb allows for up to 64 server / serial port pairs to be set for one instance of adb server that drives these emulators. 

Why 9001 ? It's just a fallback number, it's always going to be set by the initial random number that is going to be above 9001 and below 9128. 

( changing this back to 7001 is not going to change anything, the driver host port parameter is now always used, but I left the fallback port in case studio or something else still uses driver launch without the host port ? )

igorsmotto replied:
Yeah, I totally get that. 

> It's user's responsibility to make sure their ports are usable

Agreed, however, imagine that the user already took their time to setup maestro over the network, and succeeded by making sure that 7001 is a usable port. This change will now break what the user thinks is an already solved problem. 

Therefore, I propose that if sharding / parallel execution **is not being used** then 7001 should still be the default driver port. This way we avoid unnecessarily breaking the contract without, of course, the actual need to.

sdfgsdfgd replied:
of course, including this in the next  👍

sdfgsdfgd commented on maestro-cli/src/main/java/maestro/cli/App.kt:
```diff
@@ -63,10 +64,13 @@ class App {
     @Option(names = ["--host"], hidden = true)
     var host: String? = null
 
+    @Option(names = ["--driver-host-port"], hidden = true)
+    var driverHostPort: Int = Random.nextInt(9001, 9128)
```
no worries, this is a hidden parameter.

I wrote a more detailed explanation of how this makes up one of the three core parts of the patch for your question `Why the sudden change of port` . Since it's a hidden parameter, users don't have to worry about it, unless they want to specify the port themselves, as an advanced user. 

It is only an initial runtime configuration for the driver, which gets randomised so that it allows multiple maestros and multiple android drivers to run simultaneously in parallel

chris-hatton replied:
It's perhaps a fringe use case, but you _would_ need to vary the driver host port if you were wishing to use Maestro with a custom, externalised sharding solution using multiple maestro instances + emulators in the same OS, since this is the only port which would clash between the maestro's.

sdfgsdfgd replied:
🤔 with the most recent changes I'm actually randomising a not-already-used port within 9000-9128 range for each shard, as I edited my post

So there is no need to specify this any more, logic will override it in newSession anyways, I'll refactor this to be a private variable, doesn't even need to be a CLI parameter. Easy peasy

sdfgsdfgd replied:
with the most recent changes, when sharding is enabled this port is now randomised between the 9000-9128 range. 

With no sharding I'm going to keep the default global value of 7001 for old users rev compatibility, but if sharding is enabled it will allocate from between the 9000-9128 range as before.

sdfgsdfgd replied:
right, I realised this parameter variable itself is unnecessary as newSession either takes --port, or default 7001 for reverse compatibility, or if sharding is on chooses an unused port from range 7001-7128 . I removed this seperate CLI variable, the logic is already contained inside newSession or App's --port

sdfgsdfgd replied:
```
val driverHostPort = if (!sharded) parent?.port ?: 7001 else
                        (7001..7128).shuffled().find { port ->
                            usedPorts.putIfAbsent(port, true) == null
                        } ?: error("No available ports found")
```
✅

sdfgsdfgd commented:
> This PR is a big one, but most importantly is a hard one. Your ideas on how to solve this challenge are really good.
> 
> Unfortunately, let me say upfront that it will take me a while to fully review it, since of its high complexity.
> 
> In the meantime I'll be posting here all my questions as I spend more and more time in this PR.
> 
> It seems like this whole solution doesn't work on iOS::
> 
> 1. It can only create one device, probably because it tries to create them all of the same name (Error: `A device with name Maestro_iPhone11_17 is already connected`).
> 2. Even if I use already created multiple devices, it seems like only one is actually being operated by Maestro, unsure what is going on.
> 3. Haven't tested: Unsure what will happen if the user pass a list of devices that has a mix of android and devices in it. We should probably error it out, this definitely means that the user is making a mistake. This is also a weird user experience if no devices is passed, because the user is prompted multiple times what platforms it wants to run that shard.
> 
> Is this something planned? To only add support for Android? If so, then we need to error out and properly document that this feature can only work for Android
> 
> Having shards silently overrides the [sequence execution](https://maestro.mobile.dev/cli/test-suites-and-reports#sequential-execution), which can be unexpected to a user given that the sequence is a workspace-configuration. Adding support to it is non-trivial, as it would essentially means there is a need to build a graph to have some level of parallel execution, so I'd expect this to error out.
> 
> The TestCommand is responsible for creating those devices, but this suggests that the StartDeviceCommand should also have a shard parameter. This way a user can avoid having to interact with Maestro and just create a bunch of devices in a single sweep.


```
Is this something planned? To only add support for Android? If so, then we need to error out and properly document that this feature can only work for Android
```

yes, this is exactly right. Sry I forgot to mention this :) The support is initially for Android. I was hoping we could address iOS and Web support seperately in a different thread. Or unless somebody wants to contribute that here, go ahead. Otherwise I might help out with this effort and look into those after this PR is merged ? I'll be around slack as well, if you'd like to discuss more. 

```
Having shards silently overrides the [sequence execution](https://maestro.mobile.dev/cli/test-suites-and-reports#sequential-execution), which can be unexpected to a user given that the sequence is a workspace-configuration.
Adding support to it is non-trivial, as it would essentially means there is a need to build a graph to have some level of parallel execution, so I'd expect this to error out.
```
great point ! Completely missed this. I will make it error out if the sequence order thing is enabled, you're right this part would be trivial to address right now. Can get back to it at some point later perhaps. Thank u

```
The TestCommand is responsible for creating those devices, but this suggests that the StartDeviceCommand should also have a shard parameter. This way a user can avoid having to interact with Maestro and just create a bunch of devices in a single sweep.
```
agreed, addressing this as well

igorsmotto commented:
> yes, this is exactly right. Sry I forgot to mention this :) The support is initially for Android. I was hoping we could address iOS and Web support seperately in a different thread.

@sdfgsdfgd 

Got it! 

However, to not impose a worse experience to maestro users, its important for this PR to either error out if the user tries to do something that is not supported (mixing platforms or simply using iOS), or to add support for iOS. Otherwise, it will quickly turn into a bug report as a user tries to use it unknowingly.

Looking from above - I don't think this PR is that far from adding support to iOS. I see just a couple of changes that needs to be done:
- LocalXCTestInstaller is receiving the hard coded port 22087 to spawn the XCTestServer (which already receives the port to run the server by argument) inside the Simulator, it needs to be a unique port per instance. 
- To create an iOS device the name needs to be unique.
Both changes looks to be very close to exactly what you did for Android.

Let me know what you end up deciding to do!

sdfgsdfgd commented:
I think you're right, yes. iOS support seems to be a very minimal addition on top of the big work.  

I'm including the iOS support with my next commit as well, almost done with addressing this round's feedbacks.


> Looking from above - I don't think this PR is that far from adding support to iOS. I see just a couple of changes that needs to be done:
> 
> * LocalXCTestInstaller is receiving the hard coded port 22087 to spawn the XCTestServer (which already receives the port to run the server by argument) inside the Simulator, it needs to be a unique port per instance.
> * To create an iOS device the name needs to be unique.
>   Both changes looks to be very close to exactly what you did for Android.
> 
> Let me know what you end up deciding to do!

sdfgsdfgd commented:
addressed everything so far, added iOS  support for sharding. 

polished the reporting a little bit, I think verbose display of device to test mapping is going to be very helpful to all of us, see the example screenshot below

also - MAESTRO_DRIVER_STARTUP_TIMEOUT with even 2 sharded wait time easily crosses the default time of 15 secs. Maybe the default value could be increased a bit to a more realistic 20-30 secs instead. 

<img width="624" alt="Screenshot 2024-06-02 at 2 15 25 PM" src="https://github.com/mobile-dev-inc/maestro/assets/16299826/ab5d029d-602d-4dc1-bf6f-784175722137">

igorsmotto commented:
That's great @sdfgsdfgd! Thank you once again for all the hard work here! 

From my initial tests it seems like everything is working perfectly. Including iOS, which is amazing! 
I'm now in the process to run this branch against our whole internal test suite (it does take some time to run) and I'll get back to you as soon as I can.

Also, I took the liberty to push a merge commit to resolve the conflicts.

igorsmotto commented:
After testing thoroughly didn't found any regression. Thanks for the amazing work @sdfgsdfgd! 🎉 

This is probably going to be released on the next Maestro version. Release date TBD

bartekpacia commented:
See also https://github.com/mobile-dev-inc/maestro/issues/1818

:)
